### DISCUSSION TITLE
Understanding the CeleryExecutor process creation approach

### DISCUSSION THREAD

**Discussion by RonaldGalea** [apache/airflow#34341] [General] [answered]
Airflow executes every Celery task in a [separate process](https://github.com/apache/airflow/blob/main/airflow/providers/celery/executors/celery_executor_utils.py#L121), why is this the case?

For starters, Celery already can execute tasks in [separate processes](https://docs.celeryq.dev/en/stable/internals/reference/celery.concurrency.prefork.html) (which are efficiently kept in a pool) - so why create yet another brand new process for the task?

I initially thought about some custom timeout mechanism, but the parent just [blockingly](https://github.com/apache/airflow/blob/main/airflow/providers/celery/executors/celery_executor_utils.py#L131) waits for the child, so that's not it. Other than that, the current exception handling could also be done in the same process. I guess the user code could SIGKILL itself and then airflow [stats](https://github.com/apache/airflow/blob/main/airflow/providers/celery/executors/celery_executor_utils.py#L122) would not be updated, but that's quite a stretch to build new processes for.

Conversely, always creating processes, especially **from within** the celery task process is still an overhead, even with `fork()`. Consider the following DAG:

```python
with DAG(
    "tf_overhead_check",
    start_date=datetime(2021, 1, 1),
    catchup=False,
    schedule_interval=None,
) as dag:

    @task()
    def inspect_tf_overhead():
        import os
        import time

        import_start_time = time.time()
        import tensorflow as tf
        import_end_time = time.time()

        print(
            f"Import took {import_end_time - import_start_time:.4f} seconds; Process ID: {os.getpid()}; Parent process ID: {os.getppid()}"
        )

    inspection = inspect_tf_overhead()

```
Running the example above example DAG on a Celery worker takes 1+ seconds each time, because tensorflow is reimported for each task.

I emphasize **from within** because with a process pool (as Celery does), the processes are created once and could then be reused - so environment loading (at least as far as imports are concerned) is done only once.

However, if new processes are created from within celery worker pool processes (i.e. if prefork is used and tasks themselves create new processes, as currently done by Airflow), then the memory of the processes in the pool stays blank - all the loading is done in the children which exit after the task - so environments are always reloaded from scratch.

Surely I'm missing some strong reason that warrant Airflow to manually create processes for each task, but I can't figure it out?
Thank you for your insights!

**Comment by potiuk** [comment, marked as answer]
It's because of isolation. When a task makes a global change in the Python interpreter. (for example sets environment variable but there are about a million other things the process can do) we do not want side-effects to leak to the next task run by the same process in the celery worker.

**Reply by RonaldGalea** [reply]
Fair point, I suppose it's ultimately a trade-off between performance and isolation. I suppose there are no plans at the moment to leave this choice up to the user?

Also, currently Airflow allows configuring the Celery [pool](https://airflow.apache.org/docs/apache-airflow-providers-celery/stable/configurations-ref.html#pool), so one could choose a process or a thread pool. But because Airflow always creates processes, it does not actually make too much of a difference, right?

**Reply by potiuk** [reply]
Only for starting the celery master/workers. 

But if you - for example - look for an option of speeding up som imports, you have other options. I believe you could use https://airflow.apache.org/docs/apache-airflow-providers-celery/stable/configurations-ref.html#celery-config-options and specify your own configuration to load. And since it is an imported python package, you could have your custom configuration package import tensorflow in that configuration package. Then it will be imported BEFORE forking happens (either in in Celery or Airflow and  the forks will already have it imported.

Probably this is what you are looking for

**Reply by RonaldGalea** [reply]
Thanks a lot for the tip. While it would technically work, I see a number of issues:
- dependency management: I'd have to place my application dependencies system-wide for this, but I'd much prefer having a virtual env + ExternalPythonOperator.
- it only solves import related issues, not e.g. expensive to create objects (one might try to do the initialisation at import time as well, but that already gets too hacky I feel).
- potential memory issues: if child processes start modifying state (which might happen with objects from complex libraries), this might lead to lots of memory duplication (also hard to debug/inspect)

For tasks that really require warm environments, the most reasonable way I see is using Airflow as a pure orchestrator, only making `trigger_job()` and `poll_job_result()` calls to a service that in turn talks to my custom Celery cluster - as there I can run with lower levels of isolation.

**Reply by potiuk** [reply]
> * dependency management: I'd have to place my application dependencies system-wide for this, but I'd much prefer having a virtual env + ExternalPythonOperator.

This is something completely unrelated. If you want this, then your initial goal of starting quickly is thrown out of the window. You can't have both - isolated env and quick start. You seem to be contradicting yourself.

> it only solves import related issues, not e.g. expensive to create objects (one might try to do the initialisation at import time as well, but that already gets too hacky I feel).

Well you seem to want to eat cake and have it too. That calls for hacky solutions if you ask me.

> potential memory issues: if child processes start modifying state (which might happen with objects from complex libraries), this might lead to lots of memory duplication (also hard to debug/inspect)

Well. I think you pick a wrong tool for doing what you want to do. If you want multiple "units" of your work to use and reuse memory modified by a different unit, then chosing "airflow task" to be the unit is just plain wrong. You should likely use a framework that can do it, but airflow deliberately isolates tasks from each other - so that they can be even distributed - each task can run on a different machine in a remote network. Sharing memory changes between tasks is NON-FEATURE of airflow.  Posssibly you should look for something else to do it. Airflow tasks are generally supposed to be "coarse grained" and they are ONLY supposed to share any state with other tasks via XCOM. Full stop. NOT sharing state (and idempotency) is an important feature of Airflow - and the reason why many people use it.

**Reply by RonaldGalea** [reply]
I'm sorry, I probably did not formulate clearly enough.

> This is something completely unrelated. If you want this, then your initial goal of starting quickly is thrown out of the window. You can't have both - isolated env and quick start. You seem to be contradicting yourself.

In the context of Airflow you're right, with the CeleryExecutor it's either one or the other because Airflow itself is the entrypoint. This doesn't mean it's impossible in general - one could set up a lightweight Celery listener and run tasks in the same thread, Celery supports this.

> Well you seem to want to eat cake and have it too. That calls for hacky solutions if you ask me.

I'm not sure I agree. Airflow makes a decision regarding task execution, favoring isolation over start-up speed. That is completely fine and likely fits the vast majority of use cases as well as protecting users from their own potential mistakes, I can perfectly see this is a sensible approach.

However, one can also make different assumptions where looser isolation requirements are also fine. Of course, if I write a task that **does** mess up the environment and causes issues for subsequent tasks, that's on me. If my program is correct, however, there is no fundamental reason for this not to work.

> Well. I think you pick a wrong tool for doing what you want to do. If you want multiple "units" of your work to use and reuse memory modified by a different unit, then chosing "airflow task" to be the unit is just plain wrong. You should likely use a framework that can do it, but airflow deliberately isolates tasks from each other - so that they can be even distributed - each task can run on a different machine in a remote network. Sharing memory changes between tasks is NON-FEATURE of airflow. Posssibly you should look for something else to do it. Airflow tasks are generally supposed to be "coarse grained" and they are ONLY supposed to share any state with other tasks via XCOM. Full stop. NOT sharing state (and idempotency) is an important feature of Airflow - and the reason why many people use it.

Here I meant something completely different, definitely not sharing memory between tasks. I just meant that if you rely on having object X in the memory of the parent at the time of fork to also have X in the child (which is essentially your suggestion), then if the worker child modifies X, memory will have to be copied.

**Reply by potiuk** [reply]
> In the context of Airflow you're right, with the CeleryExecutor it's either one or the other because Airflow itself is the entrypoint. This doesn't mean it's impossible in general - one could set up a lightweight Celery listener and run tasks in the same thread, Celery supports this.

No. No support. This is all in the contest of those words of yours:

> dependency management: I'd have to place my application dependencies system-wide for this, but I'd much prefer having a virtual env + ExternalPythonOperator.

> I'm not sure I agree. Airflow makes a decision regarding task execution, favoring isolation over start-up speed. That is completely fine and likely fits the vast majority of use cases as well as protecting users from their own potential mistakes, I can perfectly see this is a sensible approach.

My comment was that you want to have separate dependencies, separated by Python Interpreter loading them when it starts, yet you also want the same thing warmed up. That is pretty contradicting, don't you think? Or maybe you missed the context I put it in?  Am I missing something? a

> Here I meant something completely different, definitely not sharing memory between tasks. I just meant that if you rely on having object X in the memory of the parent at the time of fork to also have X in the child (which is essentially your suggestion), then if the worker child modifies X, memory will have to be copied.

This is implicit by the way how fork works. When you have forked process it uses copy-on-write (https://en.wikipedia.org/wiki/Copy-on-write) for shared memory segments. Which means that when memory is modified in fork, the segment that contains that object gets copied and it's only modified in the fork. Which is perfect way of minimising the overhead vs. isolation.. This is guaranteed by the way how POSIX forking works - so you do not have to worry about it at all. Or maybe I am missing something? Or maybe you did not realise that ?

**Reply by RonaldGalea** [reply]
> No. No support. This is all in the contest of those words of yours

I'm just referring to how it could work if Airflow didn't interpose to run each task in a new process.

> My comment was that you want to have separate dependencies, separated by Python Interpreter loading them when it starts, yet you also want the same thing warmed up. That is pretty contradicting, don't you think? Or maybe you missed the context I put it in? Am I missing something?

No, you are completely right. My comment regarding dependencies was wrong/misplaced in this discussion.

> This is implicit by the way how fork works. When you have forked process it uses copy-on-write (https://en.wikipedia.org/wiki/Copy-on-write) for shared memory segments. Which means that when memory is modified in fork, the segment that contains that object gets copied and it's only modified in the fork. Which is perfect way of minimising the overhead vs. isolation.. This is guaranteed by the way how POSIX forking works - so you do not have to worry about it at all. Or maybe I am missing something? Or maybe you did not realise that ?

True that it's a perfect way of minimizing overhead vs isolation. But if a given task is fine with no isolation in the first place, there is no need for any trade-off. But one cannot chose this now - Airflow always makes new processes.

**Reply by potiuk** [reply]
> True that it's a perfect way of minimizing overhead vs isolation. But if a given task is fine with no isolation in the first place, there is no need for any trade-off. But one cannot chose this now - Airflow always makes new processes.

Yep. That's a deliberate choice taking into account the assumptions airflow is created with. As with everything you can have many options for the users to choose and You can even say Airflow has too many of them. There are certain assumptions (such as isolation of tasks from each other) taken  before a) we know how easy it is to leak stuff between non-isolated  tasks in an environment b) because we make a number of assumptions about task execution environment and airlfow itself is simpler because we can assume that each task is run in isolated interpreter. This makes airflow more stable, development is easiers and - most importantly time lost by maintainers on debugging weird user problems that can do strange things is miinimised. In a way we are protecting our volunteered time from those who will have strange issues because they did not realise consequences of leaking in memory data. 

So no. We will not change it, and while you might disagree with the decisions we made, I think it will stay like that. This is a very good decision we made. Again - if it's not a good decision for you, you are free to do what you want in a different way or using different tool.

**Reply by RonaldGalea** [reply]
My goal was not to contest the decisions of Airflow, but to understand them - and I can see how the design is very sensible taking all practical aspects in consideration.

I thank you again for the insightful discussions.

**Reply by potiuk** [reply]
Sure. I think knowing the context might help to make better decisions.
